pkg/dashboard/resource/stream.go

```diff
@@ -0,0 +1,168 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/nuclio/nuclio/pkg/common"
+	"github.com/nuclio/nuclio/pkg/dashboard"
+	"github.com/nuclio/nuclio/pkg/functionconfig"
+	"github.com/nuclio/nuclio/pkg/opa"
+	"github.com/nuclio/nuclio/pkg/platform"
+	"github.com/nuclio/nuclio/pkg/platform/abstract/project/external/leader/iguazio"
+	"github.com/nuclio/nuclio/pkg/platformconfig"
+	"github.com/nuclio/nuclio/pkg/restful"
+
+	"github.com/nuclio/errors"
+	"github.com/nuclio/nuclio-sdk-go"
+)
+
+type streamResource struct {
+	*resource
+}
+
+func (sr *streamResource) ExtendMiddlewares() error {
+	sr.resource.addAuthMiddleware()
+	return nil
+}
+
+func (sr *streamResource) GetAll(request *http.Request) (map[string]restful.Attributes, error) {
+
+	// ensure namespace
+	namespace := sr.getNamespaceFromRequest(request)
+	if namespace == "" {
+		return nil, nuclio.NewErrBadRequest("Namespace must exist")
+	}
+
+	// ensure project name
+	projectName := request.Header.Get("x-nuclio-project-name")
+	if projectName == "" {
+		return nil, nuclio.NewErrBadRequest("Project name must not be empty")
+	}
+
+	// get project
+	project, err := sr.getProjectByName(request, projectName, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	// get project functions
+	ctx := request.Context()
+	getFunctionsOptions := &platform.GetFunctionsOptions{
+		Name:      "",
+		Namespace: project.GetConfig().Meta.Namespace,
+		Labels: fmt.Sprintf("%s=%s",
+			common.NuclioResourceLabelKeyProjectName,
+			project.GetConfig().Meta.Name),
+		AuthSession: sr.getCtxSession(request),
+		PermissionOptions: opa.PermissionOptions{
+			MemberIds:           opa.GetUserAndGroupIdsFromAuthSession(sr.getCtxSession(request)),
+			OverrideHeaderValue: request.Header.Get(opa.OverrideHeader),
+		},
+	}
+
+	//functionsMap, _ := pr.getFunctionsAndFunctionEventsMap(request, project)
+	functions, err := sr.getPlatform().GetFunctions(ctx, getFunctionsOptions)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed getting project functions")
+	}
+
+	// iterate over functions and look for v3iostreams
+	streams, err := sr.getStreamsFromFunctions(functions)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed getting streams from functions")
+	}
+
+	return streams, nil
+}
+
+func (sr *streamResource) getStreamsFromFunctions(functions []platform.Function) (map[string]restful.Attributes, error) {
+
+	streamsMap := map[string]restful.Attributes{}
+
+	for _, function := range functions {
+		v3ioStreamsMap := functionconfig.GetTriggersByKind(function.GetConfig().Spec.Triggers, "v3ioStream")
+		for streamName, stream := range v3ioStreamsMap {
+
+			// add stream to map, with a key in the format: "function-name@stream-name"
+			keyName := fmt.Sprintf("%s@%s", function.GetConfig().Meta.Name, streamName)
```

Comment:
More of a note:
Since the project is not represented here, you indeed must have the user provide it in the header for the request
Was this something you discussed during design phase?
If we want the option to call that for all projects any time in the future, the project name will have to be represented here (before "function" level) for each entry

Reply:
This api is currently intended for the Project level view.
I am not sure the stream information has value in the outer scope level (for ALL projects) but this can be discussed further.

Reply:
No beef from me - but the current implementation of course won't hold for that, even if discussed in the future - because the project name is not represented in the structure, you'll have to break API